x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/exporter/http/HttpExportBulk.java

```diff
@@ -94,12 +91,18 @@ public void doAdd(Collection<MonitoringDoc> docs) throws ExportException {
     public void doFlush(ActionListener<Void> listener) throws ExportException {
         if (payload == null) {
             listener.onFailure(new ExportException("unable to send documents because none were loaded for export bulk [{}]", name));
-        } else if (payload.length != 0) {
+        } else if (payload.length() != 0) {
             final Request request = new Request("POST", "/_bulk");
             for (Map.Entry<String, String> param : params.entrySet()) {
                 request.addParameter(param.getKey(), param.getValue());
             }
-            request.setEntity(new NByteArrayEntity(payload, ContentType.APPLICATION_JSON));
+            try {
+                request.setEntity(new InputStreamEntity(payload.streamInput(), payload.length(), ContentType.APPLICATION_JSON));
+            } catch (IOException e) {
+                throw new AssertionError("No actual IO happens here", e);
+            }
+            // null out serialized docs to make things easier on the GC
+            payload = null;
```

Comment:
Does this really help ? It would seem that payload.streamInput() creates an object that also has a reference the same BytesReference (this). I think this line may just cause a little bit of confusion.

Reply:
I'd say, it may help a little since we're keeping a bunch of references to this doc in listeners etc. that live longer than that `InputStream` but it's hard to tell the concrete impact obviously, happy to remove it if it's too confusing :)

Reply:
not a big deal, the comment explains it well enough.